### PR TITLE
Fix issue in compile.c with a short-circuit if/unless `defined?` condition

### PR DESCRIPTION
**Pull request by luke-gru**
This caused an issue when `defined?` was in the `if` condition. Its instructions weren't appended to the instruction sequence even though it was compiled if a known logical short-circuit happened before the `defined?`. The catch table (`defined?` compilation produces a catch table) was still on the iseq even though the instructions weren't there. This caused faulty exception handling in the method. The solution is to no longer add the catch table entry when compiling `defined?` in this case.

This shouldn't touch much code, it's only for cases like the following, which can occur during debugging:

  if false && defined?(Some::CONSTANT)
    "more code..."
  end

### CODE REVIEW

**Comment by luke-gru** [comment]
I'm not sure if this is an okay fix because of language servers.

**Comment by luke-gru** [comment]
I think this change could lead to backward incompatibility with language servers, in that ruby doesn't check compilation of short-circuited logical operators now. I don't know if that's an issue or not, but if it is then there needs to be a different solution to this problem.

File: test/ruby/test_iseq.rb [hunk 1]
```diff
@@ -809,7 +809,6 @@ def test_ever_condition_loop
   def test_unreachable_syntax_error
     mesg = /Invalid break/
     assert_syntax_error("false and break", mesg)
-    assert_syntax_error("if false and break; end", mesg)
```
Comment (nobu) [inline test/ruby/test_iseq.rb:812]: Why removing this assertion?

File: bootstraptest/test_exception.rb [hunk 2]
```diff
@@ -430,3 +430,31 @@ def ===(o)
...
+      raise
+    end
+    begin
+      my_method
+    rescue
+      :ok
+    else
+      :ng
+    end
+  }, '[ruby-core:20501]'
```
Comment (nobu) [inline bootstraptest/test_exception.rb:459]: Could you move this test under test/?

File: compile.c [hunk 3]
```diff
@@ -4458,16 +4458,23 @@ static int
...
 static int
 compile_logical(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *cond,
                 LABEL *then_label, LABEL *else_label)
 {
     DECL_ANCHOR(seq);
     INIT_ANCHOR(seq);
     LABEL *label = NEW_LABEL(nd_line(cond));
-    if (!then_label) then_label = label;
-    else if (!else_label) else_label = label;
+    bool and_cond = false; bool or_cond = false;
```
Comment (nobu) [inline compile.c:4469]:
```suggestion
    bool and_cond = false, or_cond = false;
```

File: test/ruby/test_iseq.rb (hunk 1 above)
Comment (luke-gru) [reply test/ruby/test_iseq.rb:812]: Since the next part of the conditional is no longer compiled, it won't trigger that syntax error anymore. For the line just above it, it enters a different code path since there is no `if` and I didn't write this short-circuit optimization in that part of the code.

File: bootstraptest/test_exception.rb (hunk 2 above)
Comment (luke-gru) [reply bootstraptest/test_exception.rb]: done 👍

File: compile.c (hunk 3 above)
Comment (luke-gru) [reply compile.c]: 👍

File: test/ruby/test_iseq.rb (hunk 1 above)
Comment (nobu) [reply test/ruby/test_iseq.rb:812]: The `break` exists syntactically, and should be a subject of checks.

Comment (luke-gru) [reply test/ruby/test_iseq.rb:812]: If that's the case I'll try to come up with a different solution to the problem.

File: test/ruby/test_exception.rb [hunk 4]
```diff
@@ -149,6 +149,41 @@ def self.meow
...
+        rescue
+          print 'ok'
+        else
+          print 'ng'
+        end
+      }
+      # invoke in subprocess because the bug caused a segfault
+      out, _err, status = EnvUtil.invoke_ruby('--disable-gems', code, true, false)
+      assert_predicate(status, :success?, bug)
+      assert_equal 'ok', out
```
Comment (nobu) [inline test/ruby/test_exception.rb:183]: We have the dedicated method to run assertions in a separated process.

```suggestion
      assert_separately([], "#{<<~"begin;"}\n#{<<~'end;'}")
      begin;
        def my_method
          var = nil
          if #{cond}
            var = var = "here"
          end
          raise
        end
        assert_raise(RuntimeError) do
          my_method
        end
      end;
```

And this doesn't seem related to behaviors of exception, maybe test_syntax.rb?

**Review by nobu — CHANGES_REQUESTED** [review]

File: test/ruby/test_exception.rb [hunk 5]
```diff
@@ -149,6 +149,41 @@ def self.meow
     end
   end
 
+  def test_exception_with_defined_in_short_circuit_if_condition
+    bug = '[ruby-core:20501]'
+    conds = [
+      "false && defined?(Some::CONSTANT)",
+      "true || defined?(Some::CONSTANT)",
+      "(false && defined?(Some::CONSTANT))",
+      "(true || defined?(Some::CONSTANT))",
```
Comment (nobu) [inline test/ruby/test_exception.rb:158]: These lines are same as previous lines, except for parentheses.
Necessary?

Comment (luke-gru) [reply test/ruby/test_exception.rb]: It exercises a different code path, I put a comment about that in my most recent commit on this PR

File: test/ruby/test_exception.rb (hunk 4 above)
Comment (luke-gru) [reply test/ruby/test_exception.rb]: Ok I moved to test_syntax, but I just tried using assert_separately and couldn't reproduce the issue. I could reproduce it only in a clean environment with no test environment required, so that's why I didn't use it. I've added a comment about that in my most recent commit.

File: tool/lib/envutil.rb [hunk 6]
```diff
@@ -165,6 +165,11 @@ def invoke_ruby(args, stdin_data = "", capture_stdout = false, capture_stderr =
     }
 
     args = [args] if args.kind_of?(String)
+    # use the same parser as current ruby
+    if args.none? { |arg| arg.to_s =~ /--parser=/ }
+      current_parser = RUBY_DESCRIPTION =~ /prism/i ? "prism" : "parse.y"
+      args << "--parser=#{current_parser}"
+    end
```
Comment (luke-gru) [inline tool/lib/envutil.rb:172]: I need these lines because prism is the current parser now (it wasn't when I made my initial commit). So now it only used prism parser and compiler in a subprocess without this. Now, when parse.y is tested it will use parse.y parser/compiler in subprocess too.

File: tool/lib/envutil.rb [hunk 7]
```diff
@@ -165,6 +165,11 @@ def invoke_ruby(args, stdin_data = "", capture_stdout = false, capture_stderr =
     }
 
     args = [args] if args.kind_of?(String)
+    # use the same parser as current ruby
+    if args.none? { |arg| arg.to_s =~ /--parser=/ }
```
Comment (nobu) [inline tool/lib/envutil.rb:169]:
```suggestion
    if args.none? { |arg| arg.start_with?("--parser=") }
```

**Review by nobu — APPROVED** [review]
